sql: stricter type checking on GREATEST and LEAST

madelynnblue opened the pull request:
Change the returnType checking to return a func that can dynamically
type check based on the arguments it receives. This is required
because GREATEST and LEAST can return either a float or an int, based
on their arguments. Before this change, lib/pq would panic under some
conditions listed in #3294.

To prevent this, we need to make sure all the arguments to GREATEST
and LEAST are of the same type. We could perform this check in the
implementation of GREATEST and LEAST themselves (which would trigger
an error in the type checker), but this requires running the check
on all executions of those functions. Instead, we can perform it only
once during the type checking phase.

Fixes #3294.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3298)

<!-- Reviewable:end -->

tamird commented:
Nit: in the commit message s/fixes #3294/Fixes #3294./

LGTM. Nice work.

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[sql/parser/builtins.go, line 1257 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298#-K4Zs-SfJWL1K4tRMXmH:-K4Zs-SgUV-wYAtShz3F:-217205909) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb9449e98394e76c0142b869d50d514025cd43e4/sql/parser/builtins.go#L1257)):*
s/return/type/ to match other functions?

---

*[sql/parser/builtins.go, line 1258 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298#-K4ZsZMGOZ_eLuKWcGL9:-K4ZsZMHdM0jgSdGeAj8:-16091798) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb9449e98394e76c0142b869d50d514025cd43e4/sql/parser/builtins.go#L1258)):*
using a string is not the way to go; instead of calling Type(), call TypeCheck() and you'll get back a datum that you can compare directly

also, this variant of the loop uses a style more in line with existing code:
```go
var dummy Datum
for _, dummyArg := range args {
  if dummy == DNull {
    dummy = dummyArg
    continue
  }
  if !(dummy == DNull || dummy == dummyArg) {
    return nil, fmt.Errorf("incompatible argument types %s, %s", dummy.Type(), dummyArg.Type())
  }
}
return dummy, nil
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298#-:-K4ZrRGfMUvllx1xfwoa:-1738101660)*
<!-- Sent from Reviewable.io -->

tschottdorf commented on sql/parser/builtins.go:
```diff
@@ -1250,6 +1252,28 @@ func round(x float64, n int64) (Datum, error) {
+		}
+	}
+	for _, a := range args[i:] {
+		if a.Type() != typ && a != DNull {
```
ok if they're all `NULL`?

Also, super nit: you're comparing `args[i]` with itself (but I don't think avoiding that is net positive).

petermattis commented on sql/parser/type_check.go:
```diff
@@ -25,6 +25,16 @@ import (
 )
 
+var (
+	TypeBytes     = func(DTuple) (Datum, error) { return DummyBytes, nil }
```
Do these variables need to be exported?

petermattis commented on sql/parser/builtins.go:
```diff
@@ -1250,6 +1252,28 @@ func round(x float64, n int64) (Datum, error) {
+	for ; i < len(args); i++ {
+		arg = args[i]
+		if arg != DNull {
+			typ = arg.Type()
```
Better to use `reflect.TypeOf(arg)` than string comparison.

petermattis commented:
LGTM

mjibson commented:
Review status: 1 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/parser/builtins.go, line 1257 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298#-K4Zs-SfJWL1K4tRMXmH:-K4aOSlvgGbtn-E22fMp:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb9449e98394e76c0142b869d50d514025cd43e4/sql/parser/builtins.go#L1257)):*
Done.

---

*[sql/parser/builtins.go, line 1258 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298#-K4ZsZMGOZ_eLuKWcGL9:-K4aQ9fzh5appihtgfUe:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb9449e98394e76c0142b869d50d514025cd43e4/sql/parser/builtins.go#L1258)):*
Done.

---

*[sql/parser/builtins.go, line 1265 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298#-K4ZqhqRYysnN0_En8ve-r1-1265:-K4aQA5QjmQr5RXjtEEA:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb9449e98394e76c0142b869d50d514025cd43e4/sql/parser/builtins.go#L1265)):*
Done.

---

*[sql/parser/builtins.go, line 1270 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298#-K4ZqhqRYysnN0_En8ve-r1-1270:-K4aQBMgV6YcujEymHns:507933655) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb9449e98394e76c0142b869d50d514025cd43e4/sql/parser/builtins.go#L1270)):*
I refactored the loop to avoid this problem. And since datum starts as DNull, that's what will be returned if all arguments are null.

---

*[sql/parser/type_check.go, line 29 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298#-K4ZqhqS15rUFyd8Il_f-r1-29:-K4aRB9CvtiuAIUGPq1Z:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb9449e98394e76c0142b869d50d514025cd43e4/sql/parser/type_check.go#L29)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 3 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sql/parser/builtins.go, line 1258 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298#-K4ZsZMGOZ_eLuKWcGL9:-K4awuNiEW3sSBCcTXj7:-39118374) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb9449e98394e76c0142b869d50d514025cd43e4/sql/parser/builtins.go#L1258)):*
Why not provide the better error message shown above?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 3 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[sql/testdata/builtin_function, line 782 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298#-K4bpTl2w3aeCbpAsdR8:-K4bpTl2w3aeCbpAsdR9:-637203766) ([raw file](https://github.com/cockroachdb/cockroach/blob/503413cf1937dfd376ca0f68d038bfec2e7b8114/sql/testdata/builtin_function#L782)):*
i'm pretty sure your implentation currently has a bug. add this test case `SELECT GREATEST(1, NULL)`; i believe it will spuriously error

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 1 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/parser/builtins.go, line 1258 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298#-K4ZsZMGOZ_eLuKWcGL9:-K4cpJ5ok_gvPk4KMr7C:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb9449e98394e76c0142b869d50d514025cd43e4/sql/parser/builtins.go#L1258)):*
Done.

---

*[sql/testdata/builtin_function, line 782 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298#-K4bpTl2w3aeCbpAsdR8:-K4cnD-nWWAhdiv5ppYA:-1825572378) ([raw file](https://github.com/cockroachdb/cockroach/blob/503413cf1937dfd376ca0f68d038bfec2e7b8114/sql/testdata/builtin_function#L782)):*
No, this works fine.  I tested your case here specifically and it worked. The tests already have pretty much that anyway:

```
query I
SELECT GREATEST(NULL, NULL, NULL, NULL, 2)
----
2

query I
SELECT GREATEST(2, NULL, NULL, NULL, NULL)
----
2
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sql/parser/builtins.go, line 1258 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298#-K4ZsZMGOZ_eLuKWcGL9:-K4cuhdQXkXrww-l3R8_:1609111590) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb9449e98394e76c0142b869d50d514025cd43e4/sql/parser/builtins.go#L1258)):*
OK, so now you're doing almost what I originally suggested, except with additional calls to `reflect.TypeOf`. Why don't you compare the data directly like I suggested?

---

*[sql/testdata/builtin_function, line 782 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298#-K4bpTl2w3aeCbpAsdR8:-K4cvE5RZmNTEFdxIGQY:-265821638) ([raw file](https://github.com/cockroachdb/cockroach/blob/503413cf1937dfd376ca0f68d038bfec2e7b8114/sql/testdata/builtin_function#L782)):*
Oh, I missed the `if arg == DNull { continue }`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sql: stricter type checking on GREATEST and LEAST

Change the returnType checking to return a func that can dynamically
type check based on the arguments it receives. This is required
because GREATEST and LEAST can return either a float or an int, based
on their arguments. Before this change, lib/pq would panic under some
conditions listed in #3294.

To prevent this, we need to make sure all the arguments to GREATEST
and LEAST are of the same type. We could perform this check in the
implementation of GREATEST and LEAST themselves (which would trigger
an error in the type checker), but this requires running the check
on all executions of those functions. Instead, we can perform it only
once during the type checking phase.

Fixes #3294

mjibson commented:
Review status: 1 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/parser/builtins.go, line 1258 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298#-K4ZsZMGOZ_eLuKWcGL9:-K4d3pok0LmlWYJMzRbc:-73760153) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb9449e98394e76c0142b869d50d514025cd43e4/sql/parser/builtins.go#L1258)):*
I was using `reflect.TypeOf` because that's what Peter suggested to use (see the comments below this set). I wasn't comparing the data directly because I didn't think it would work since `DInt(0) != DInt(1)`. But, since this is called with dummy arguments that are fetched via other results of `TypeCheck()`, which all boil down to the Dummy vars, equality can be used.

Has been updated to just use direct comparison.

---

*[sql/testdata/builtin_function, line 782 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298#-K4bpTl2w3aeCbpAsdR8:-K4d42oJjD-jxI3pKjTK:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/503413cf1937dfd376ca0f68d038bfec2e7b8114/sql/testdata/builtin_function#L782)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3298)*
<!-- Sent from Reviewable.io -->
